rust/src/lockfile.rs

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright (C) 2018 Red Hat, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+use c_utf8::CUtf8Buf;
+use serde_derive::{Deserialize, Serialize};
+use serde_json;
+use serde_yaml;
+use std::collections::HashMap;
+use std::path::Path;
+use std::{fs, io};
+use failure::Fallible;
+use failure::ResultExt;
+
+pub struct Lockfile {
+    #[allow(dead_code)] // Not used in tests
+    parsed: LockfileConfig,
+    serialized: CUtf8Buf
+}
+
+#[derive(PartialEq)]
+enum InputFormat {
+    YAML,
+    JSON,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct LockfileConfig {
+    packages: HashMap<String, Vec<(String, String)>>
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct PermissiveLockfileConfig {
+    #[serde(flatten)]
+    config: LockfileConfig,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+struct StrictLockfileConfig {
```

Comment:
No reason to have permissive/strict here; that's a legacy holdover from the treefile code.  We should always be strict.